### PR TITLE
CSE-1667 rest api endpoint

### PR DESCRIPTION
**Pull request by dominikstoetter**
**Jira Issue**

https://8select.atlassian.net/browse/CSE-1667

**Summary**

- add api endpoints
```
/cse-eightselect-basic/products?format=product_feed
/cse-eightselect-basic/products?format=property_feed
/cse-eightselect-basic/products?format=status_feed
```
- authorization fails and endpoint will be hidden if:
-- `8select-com-tid` is missing
-- `8select-com-fid` is missing
- if validation fails because of credential missmatch endpoint throws 403
- if validation fails because of unconfigured plugin endpoint throws 500
- return 200 with dummy body if request is valid

**Checklist**

- [x] PR title is prefixed with Jira Issue Id - e.g. CSE-42
- [x] Add feature / bug label
- [ ] Unit tests for new / changed logic exist OR no logic changes are passing

bildschirmfoto 2018-12-21 um 08 39 42

### CODE REVIEW

**Comment by akleiber** [comment]
please use one plural ressource and specify the type via type - the only real difference between the 3 type is the format - the contents is the same

you get:

- restful naming the follow our patterns we use in all our rest apis
- less code because you only use one controller action that just needs to inject the type, limit, offset into the export class
```
/cse-eightselect-basic/products?type=product
/cse-eightselect-basic/products?type=property
/cse-eightselect-basic/products?type=status
```

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php [hunk 1]
```diff
@@ -23,4 +25,39 @@ public function cartAction()
...
+    public function productsAction() {
+        try {
+            $container = Shopware()->Container();
+            $this->Front()->Plugins()->ViewRenderer()->setNoRender();
+
+            $export = $container->get('cse_eightselect_basic.export.export');
+            $requestWasValid = $export->validateExportRequest($this->Response(), $this->Request());
+
+            if (!$requestWasValid) {
+                return false;
```
Comment (voodoods) [inline CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: do we wan't to log or return something if the request wasn't valid?

File: CseEightselectBasic/Services/Export/Export.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    $feedId = $request->getHeader('8select-com-fid');
+    
+    return $tenantId != null && $feedId != null;
+  }
+
+  private function isValidFormatSpecified ($request) {
+    $format = self::getFormatParam($request);
+    
+    if ($format == null) {
+      return false;
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: do we wan't to return an error message that the specified format was not valid?

File: CseEightselectBasic/Services/Export/Export.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private function isTidAndFidValid ($request) {
+    $tenantIdMatches = $request->getHeader('8select-com-tid') === $this->pluginConfig['CseEightselectBasicApiId'];
+    $feedIdMatches = $request->getHeader('8select-com-fid') === $this->pluginConfig['CseEightselectBasicFeedId'];
+    return $tenantIdMatches && $feedIdMatches;
+  }
+
+  public function getOffsetAndLimit ($request) {
+    //@todo maybe use symfony
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: remove this comment if not needed pls

File: CseEightselectBasic/Services/Export/Export.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    if(!is_numeric($limit) || !is_numeric($offset)) {
+      return false;
+    }
+
+    $limit = intval($limit);
+    $offset = intval($offset);
+  
+    if($limit > $MAX_LIMIT) {
+      return false;
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: do we want to return an error message that the provided limit is bigger than the max. allowed limit?

File: CseEightselectBasic/Services/Export/Export.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    return ['limit' => $limit, 'offset' => $offset];
+  }
+
+  public function validateExportRequest($response, $request) {
+    $isTidAndFidInRequest = self::isTidAndFidInRequest($request);
+    $isValidFormatSpecified = self::isValidFormatSpecified($request);
+
+    if (!$isTidAndFidInRequest || !$isValidFormatSpecified) {
+      $response->setHttpResponseCode(404);
+      return false;
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: do we want to return an error message telling the requester that `tid` and `fid` or the given `format` aren't provided?

File: CseEightselectBasic/Services/Export/Export.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        $response->setHeader('Content-Type', 'application/json');
+        $response->setBody(json_encode($validationResult['violations']));
+        $response->setHttpResponseCode(500);
+        return false;
+    }
+    
+    $areTidAndFidValid = self::isTidAndFidValid($request);
+    if (!$areTidAndFidValid) {
+        $response->setHeader('Content-Type', 'text/html');
+        $response->setBody('wrong tid or fid');
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: good one, but don't the config validation results already contain something for that?

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php (hunk 1 above)
Comment (dominikstoetter) [reply CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: response happens in the Export.php :) this is just an abort

File: CseEightselectBasic/Services/Export/Export.php (hunk 6 above)
Comment (dominikstoetter) [reply CseEightselectBasic/Services/Export/Export.php]: no they just check if api-id and feed-id are set in the configuration. This one checks for the actual validity of them

File: CseEightselectBasic/Services/Export/Export.php (hunk 5 above)
Comment (dominikstoetter) [reply CseEightselectBasic/Services/Export/Export.php]: no we explicitly want to hide the api in those cases

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php [hunk 7]
```diff
@@ -23,4 +25,39 @@ public function cartAction()
...
+
+    /**
+     * The API is available at /eight-select/exportProduct
+     */
+    public function productsAction() {
+        try {
+            $container = Shopware()->Container();
+            $this->Front()->Plugins()->ViewRenderer()->setNoRender();
+
+            $export = $container->get('cse_eightselect_basic.export.export');
```
Comment (voodoods) [inline CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: rename this class to validator and create another one for the export logic

File: CseEightselectBasic/Services/Export/Export.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    if (!$offsetAndLimit) {
+        $response->setHeader('Content-Type', 'text/html');
+        $response->setBody('offset or limit missing or no integer');
+        $response->setHttpResponseCode(400);
+        return false;
+    }
+    return true;
+  } 
+
+  public function getArticlesByOffsetAndLimit($offsetAndLimit) {
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: remove this method and create a class for the export. Single responsibility!

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php (hunk 7 above)
Comment (voodoods) [inline CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: $this->container->get('...')

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php [hunk 9]
```diff
@@ -23,4 +25,39 @@ public function cartAction()
...
+            $this->Response()->setHttpResponseCode(500);
+        }
+    }
+
+    public function getWhitelistedCSRFActions() {
+		return [
+			'all',
+			'delta'
+		];
+	}
```
Comment (voodoods) [inline CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: prettify

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php [hunk 10]
```diff
@@ -23,4 +25,39 @@ public function cartAction()
...
+            $this->Response()->setHeader('Content-Type', 'text/html');
+            $this->Response()->setBody($e->getMessage());
+            $this->Response()->setHttpResponseCode(500);
+        }
+    }
+
+    public function getWhitelistedCSRFActions() {
+		return [
+			'all',
+			'delta'
```
Comment (voodoods) [inline CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: these needs to be actions that are going to be whitelisted. in this case products
https://developers.shopware.com/developers-guide/csrf-protection/#whitelist-particular-actions

File: CseEightselectBasic/Services/Export/Export.php [hunk 11]
```diff
@@ -0,0 +1,105 @@
+<?php
+namespace CseEightselectBasic\Services\Export;
+
+use Shopware;
+
+class Export {
+  public function __construct() {
+    $container = Shopware()->Container();
+    $this->configValidator = $container->get('cse_eightselect_basic.config.validator');
+    $this->pluginConfig = $container->get('cse_eightselect_basic.dependencies.provider')->getPluginConfig();
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: inject needed services via services.xml arguments, 

see documantation for further usage
https://developers.shopware.com/blog/2015/11/11/best-practices-of-shopware-plugin-development/#dependency-injection

File: CseEightselectBasic/Services/Export/Export.php [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    $offset = intval($offset);
+  
+    if($limit > $MAX_LIMIT) {
+      return false;
+    }
+
+    return ['limit' => $limit, 'offset' => $offset];
+  }
+
+  public function validateExportRequest($response, $request) {
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: add docblocks and type the arguments

File: CseEightselectBasic/Services/Export/Export.php [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    $isTidAndFidInRequest = self::isTidAndFidInRequest($request);
+    $isValidFormatSpecified = self::isValidFormatSpecified($request);
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: self is only for static methods.

File: CseEightselectBasic/Services/Export/Export.php [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        $response->setBody('wrong tid or fid');
+        $response->setHttpResponseCode(400);
+        return false;
+    }
+    
+    $offsetAndLimit = self::getOffsetAndLimit($request);
+    if (!$offsetAndLimit) {
+        $response->setHeader('Content-Type', 'text/html');
+        $response->setBody('offset or limit missing or no integer');
+        $response->setHttpResponseCode(400);
```
Comment (voodoods) [inline CseEightselectBasic/Services/Export/Export.php]: always return json https://gist.github.com/akleiber/32b2d1a4d9fa6ad8bf95f1a6488fb143#fehler

File: CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php (hunk 9 above)
Comment (dominikstoetter) [reply CseEightselectBasic/Controllers/Frontend/CseEightselectBasic.php]: bildschirmfoto 2018-12-18 um 16 49 15
i have no clue what github does here

**Comment by akleiber** [comment]
```
curl  http://shopware-5-5-3-php7.dev.8select/cse-eightselect-basic/products\?limit=5 -H '8select-com-tid:8cb71563-c3c7-4bbe-afe5-2c236478ce29' -H '8select-com-fid:8cb71563-c3c7-4bbe-afe5-2c236478ce29'
```

```
{"limit":"5","offset":0,"total":580,"data":[{"sku":"SW42"},{"sku":"SW43"},{"sku":"SW44"}]}
```

**Pull request merged**